src/workerd/api/sockets.h

```diff
@@ -18,16 +18,27 @@ struct SocketAddress {
 
 struct SocketOptions {
   jsg::Unimplemented tls; // TODO(later): TCP socket options need to be implemented.
-  JSG_STRUCT(tls);
+  bool allowHalfOpen = false;
+  JSG_STRUCT(tls, allowHalfOpen);
 };
 
 class Socket: public jsg::Object {
 public:
-  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readable, jsg::Ref<WritableStream> writable,
-      kj::Own<jsg::PromiseResolverPair<void>> close)
-      : readable(kj::mv(readable)), writable(kj::mv(writable)),
+  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readableParam, jsg::Ref<WritableStream> writable,
+      kj::Own<jsg::PromiseResolverPair<void>> close, kj::Promise<void> connDisconnPromise,
+      jsg::Optional<SocketOptions> options)
+      : readable(kj::mv(readableParam)), writable(kj::mv(writable)),
         closeFulfiller(kj::mv(close)),
-        closedPromise(kj::mv(closeFulfiller->promise)) { };
+        closedPromise(kj::mv(closeFulfiller->promise)),
+        writeDisconnectedPromise(nullptr),
+        options(kj::mv(options)) {
+    // Listen for abrupt disconnects and resolve the `closed` promise when they occur.
+    auto& ioContext = IoContext::current();
+    writeDisconnectedPromise = ioContext.addObject(
+        kj::heap(ioContext.awaitIo(kj::mv(connDisconnPromise)).then(js, [this](jsg::Lock& js) {
+          closeFulfiller->resolver.resolve();
+        })));
```

Comment:
Nit: Just move this up into the constructor init list... e.g. instead of `writeDisconnectedPromise(nullptr)` ... make it `writeDisconnectedPromise(IoContext::current().addObject...)`